### PR TITLE
Removed reference to the masterCredentialsRef in the tenant CRD and use the APIManager namespace instead

### PR DESCRIPTION
**Pull request by austincunningham**
# What
Removed reference to the masterCredentialsRef in the tenant CRD and use the APIManager namespace instead

# Verification
- deploy this branch locally in the a `3scale-test` namespace
- Create an apimanager
- create a secret in the apimanager namespace
image
- create a tenant CR with the follow spec
```yaml
spec:
  apiManagerNamespaceRef: 3scale-test
  email: tenant1@emails.com
  organizationName: tenant1-org
  passwordCredentialsRef:
    name: tenant-5-admin-password
    namespace: 3scale-test
  systemMasterUrl: '<master url e.g. https://master.apps.aucunnin.egcw.s1.devshift.org/'>
  tenantSecretRef:
    name: provider-account-ref
    namespace: 3scale-test
  username: tenant1
```
- After the tenant reconciles you should see the status update e.g.
```yaml
status:
  adminId: 5
  tenantId: 4
```

### CODE REVIEW

**Comment by austincunningham** [comment]
@eguzki if you have time for a review

**Comment by austincunningham** [comment]
@MStokluska @KevFan might be good if you can review this also

File: apis/capabilities/v1alpha1/tenant_types.go [hunk 1]
```diff
@@ -39,7 +41,8 @@ type TenantSpec struct {
 	SystemMasterUrl        string             `json:"systemMasterUrl"`
 	TenantSecretRef        v1.SecretReference `json:"tenantSecretRef"`
 	PasswordCredentialsRef v1.SecretReference `json:"passwordCredentialsRef"`
-	MasterCredentialsRef   v1.SecretReference `json:"masterCredentialsRef"`
+	// Namespace of the existing APIManager that you wish to used to create the Tenant
+	APIManagerNamespaceRef string `json:"apiManagerNamespaceRef"`
```
Comment (eguzki) [inline apis/capabilities/v1alpha1/tenant_types.go:45]: This is what makes me feel uneasy. We are breaking backward compat. From the product's point of view it is ok, as this is still in tech preview. However, if there is a tenant CR created in a cluster and the operator is upgraded to a new version with a new Tenant CRD, what happens?

**Review by eguzki — CHANGES_REQUESTED** [review]
The code changes loook ok. 

Missing doc refactor in https://github.com/3scale/3scale-operator/blob/master/doc/operator-application-capabilities.md#tenant-custom-resource

File: apis/capabilities/v1alpha1/tenant_types.go (hunk 1 above)
Comment (austincunningham) [reply apis/capabilities/v1alpha1/tenant_types.go:45]: We sort of knew there would be breaking changes. Should we add back in MasterCredentialsRef and make it optional and handle the logic with either scenario and update the docs to use APIManagerNamespaceRef only?

**Comment by eguzki** [comment]
you need to update bundles as well: `make bundle`

**Comment by eguzki** [comment]
I have tested the upgrade when there is a tenant CR existing in the cluster. 

TL;DR: The upgrade fails and operator does not upgrade.

This was somewhat expected. 

In the UI, the dashboard shows upgrade pending

Screenshot 2022-04-13 at 18-38-34 Installed Operators · Red Hat OpenShift Container Platform

And the install plan gives a clear message about the error. 

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: InstallPlan
metadata:
  creationTimestamp: "2022-04-13T16:25:21Z"
  generateName: install-
  generation: 1
  labels:
    operators.coreos.com/3scale-operator.3scale: ""
  name: install-qdch8
  namespace: 3scale
  ownerReferences:
    - apiVersion: operators.coreos.com/v1alpha1
      blockOwnerDeletion: false
      controller: false
      kind: Subscription
      name: 3scale-operator
      uid: b39ea05c-a7c0-45fd-accf-ab850771c3f4
  resourceVersion: "49237805"
  uid: 817bc5db-89a3-41da-8fa8-a77c2f377199
spec:
  approval: Automatic
  approved: true
  clusterServiceVersionNames:
    - 3scale-operator.v0.9.0
  generation: 4
status:
  bundleLookups:
    - catalogSourceRef:
        name: eguzki-catalog
        namespace: openshift-marketplace
      identifier: 3scale-operator.v0.9.0
      path: quay.io/3scale/3scale-operator:eguzki-2.12-operator-bundle3
      properties: '{"properties":[{"type":"olm.gvk","value":{"group":"apps.3scale.net","kind":"APIManagerRestore","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"apps.3scale.net","kind":"APIManager","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"ActiveDoc","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"Backend","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"DeveloperAccount","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"DeveloperUser","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"Product","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"apps.3scale.net","kind":"APIManagerBackup","version":"v1alpha1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"CustomPolicyDefinition","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"OpenAPI","version":"v1beta1"}},{"type":"olm.gvk","value":{"group":"capabilities.3scale.net","kind":"Tenant","version":"v1alpha1"}},{"type":"olm.package","value":{"packageName":"3scale-operator","version":"0.9.0"}}]}'
      replaces: 3scale-operator.v0.8.3-0.1646742992.p
  catalogSources: []
  conditions:
    - lastTransitionTime: "2022-04-13T16:26:46Z"
      lastUpdateTime: "2022-04-13T16:26:46Z"
      message: 'error validating existing CRs against new CRD''s schema for "tenants.capabilities.3scale.net": error validating custom resource against new schema for Tenant 3scale/ecorp-tenant: [].spec.apiManagerNamespaceRef: Required value'
      reason: InstallComponentFailed
      status: "False"
      type: Installed
  message: 'error validating existing CRs against new CRD''s schema for "tenants.capabilities.3scale.net": error validating custom resource against new schema for Tenant 3scale/ecorp-tenant: [].spec.apiManagerNamespaceRef: Required value'
  phase: Failed
```

So, options we have as I see:
* Go with this change and add documentation note about deleting all the tenant CRs first before upgrading
* Create a new version in the CRD and Implement some conversion webhook. https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/
* Discard the new `apiManagerNamespaceRef` field

Do you see any other feasible option?

IMHO, this new `apiManagerNamespaceRef` provides a slightly better UX, but the security issue is still there. So, I tend to think this extra work effort required does not pay off.

**Comment by austincunningham** [comment]
@eguzki I tend to agree `apiManagerNamespaceRef` does look better but doesn't improve the initial issue in any way . I don't think its worth the effort as it breaks the upgrade path.  

cc @mkudlej

**Comment by austincunningham** [comment]
closing this pr
